### PR TITLE
Add tracing_attributes to tracing decorator

### PR DESCRIPTION
**Pull request by lmazuel**
Expected usage:
```python
storage_distributed_trace = functool.partial(distributed_trace, tracing_attributes={'az.namespace', 'MicrosoftStorage'})

@storage_distributed_trace
def upload_blob():
   ...
```

Step in https://github.com/Azure/azure-sdk-for-python/issues/9256

### CODE REVIEW

**Comment by lmazuel** [comment]
@johanste any thoughts?

File: sdk/core/azure-core/tests/azure_core_asynctests/test_tracing_decorator_async.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        with FakeSpan(name="parent") as parent:
+            client = MockClient(policies=[])
+            await client.get_foo(parent_span=parent)
+            await client.get_foo()
+
+            assert len(parent.children) == 3
+            assert parent.children[0].name == "MockClient.__init__"
+            assert not parent.children[0].children
+            assert parent.children[1].name == "MockClient.get_foo"
+            assert not parent.children[1].children
```
Comment (bryevdv) [inline sdk/core/azure-core/tests/azure_core_asynctests/test_tracing_decorator_async.py]: no assertions for third child?

File: sdk/core/azure-core/tests/test_tracing_decorator.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        with FakeSpan(name="parent") as parent:
+            client = MockClient(policies=[])
+            client.get_foo(parent_span=parent)
+            client.get_foo()
+
+        assert len(parent.children) == 3
+        assert parent.children[0].name == "MockClient.__init__"
+        assert not parent.children[0].children
+        assert parent.children[1].name == "MockClient.get_foo"
+        assert not parent.children[1].children
```
Comment (bryevdv) [inline sdk/core/azure-core/tests/test_tracing_decorator.py]: same comment re: third child

File: sdk/core/azure-core/tests/test_tracing_decorator.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                client.raising_exception()
+            except:
+                pass
+            client.get_foo()
+
+        assert len(parent.children) == 3
+        assert parent.children[0].name == "MockClient.__init__"
+        assert parent.children[1].name == "MockClient.raising_exception"
+        # Exception should propagate status for Opencensus
+        assert parent.children[1].status == 'Something went horribly wrong here'
```
Comment (bryevdv) [inline sdk/core/azure-core/tests/test_tracing_decorator.py]: Does `.status` have a value that can be maintained in successful cases?

Comment (lmazuel) [reply sdk/core/azure-core/tests/test_tracing_decorator.py]: That's a good question, setting status is actually the resability of the plugin implementation (which are not in this PR). That being said, both OpenCensus and OpenTelemetry does nothing with status, until there is an exception when they make sure status is called. This test is mostly to check that it's possible to do that (the type of status in a real implementation doesn't have to be a string).
TLDR: I don't think it makes sense to check status on non-error scenario here (but we could in plugin implementation)

File: sdk/core/azure-core/tests/azure_core_asynctests/test_tracing_decorator_async.py (hunk 1 above)
Comment (lmazuel) [reply sdk/core/azure-core/tests/azure_core_asynctests/test_tracing_decorator_async.py]: he's grounded :)
we could, but that's not the point of this test, doesn't hurt to add it though

Comment (lmazuel) [reply sdk/core/azure-core/tests/azure_core_asynctests/test_tracing_decorator_async.py]: fixed

File: sdk/core/azure-core/tests/test_tracing_decorator.py (hunk 2 above)
Comment (lmazuel) [reply sdk/core/azure-core/tests/test_tracing_decorator.py]: fixed

**Comment by bryevdv** [comment]
LGTM (assuming will be green with latest commit :) )

File: sdk/core/azure-core/azure/core/tracing/decorator.py [hunk 4]
```diff
@@ -36,38 +36,48 @@
     TYPE_CHECKING = False
 
 if TYPE_CHECKING:
-    from typing import Callable, Any
+    from typing import Callable, Dict, Optional, Any, cast
 
 
-def distributed_trace(func=None, name_of_span=None):
-    # type: (Callable, str) -> Callable[[Any], Any]
+def distributed_trace(_func=None, name_of_span=None, tracing_attributes=None):
```
Comment (johanste) [inline sdk/core/azure-core/azure/core/tracing/decorator.py]: keyword only argument? I don't expect there to be any logical ordering of any additional parameters we may add in the future...

Comment (lmazuel) [reply sdk/core/azure-core/azure/core/tracing/decorator.py]: Fixed
